### PR TITLE
add receive to uart

### PR DESCRIPTION
**Pull request by yevgenfs**
add receive to uart

### CODE REVIEW

File: Src/lib/UL/uart_manager/uart_manager.c [hunk 1]
```diff
@@ -34,22 +37,22 @@ e_uart_manager_err_t uart_manager_deinit()
...
 }
 
 e_uart_manager_err_t convert_payload(led_ctrl_t led_state)
 {
-    package.payload.led_status = led_state;
+    send_package.send_payload.led_status = led_state;
     return e_uart_manager_err_ok;
 }
 
 e_uart_manager_err_t convert_message()
```
Comment (oleh-plotnikov) [inline Src/lib/UL/uart_manager/uart_manager.c]: Use input argument here instead send_package.send_header.led_num, send_package.send_payload.led_status

File: Src/lib/UL/uart_manager/uart_manager.c [hunk 2]
```diff
@@ -70,9 +73,72 @@ e_uart_manager_err_t send_to_uart(led_num_t led_num, led_ctrl_t led_state)
...
 
-    if (uart_send(&uart, message) != e_uart_err_ok)
+    if (uart_send(&uart, send_message) != e_uart_err_ok)
     {
         return e_uart_manager_err_invalid_argument;
     }
     return e_uart_manager_err_ok;
 }
+
+e_uart_manager_err_t convert_receive_package(uint8_t *receive_message)
```
Comment (oleh-plotnikov) [inline Src/lib/UL/uart_manager/uart_manager.c]: You should send receive_message length as input arg also

File: Src/stm32f4xx_it.c [hunk 3]
```diff
@@ -197,6 +197,26 @@ void SysTick_Handler(void)
 /* please refer to the startup file (startup_stm32f4xx.s).                    */
 /******************************************************************************/
 
+/**
+  * @brief This function handles USART2 global interrupt.
+  */
+void USART2_IRQHandler(void)
+{
+  /* USER CODE BEGIN USART2_IRQn 0 */
```
Comment (oleh-plotnikov) [inline Src/stm32f4xx_it.c:205]: Not correct interrupt handling

**Review by oleh-plotnikov — CHANGES_REQUESTED** [review]

File: Src/lib/UL/uart_manager/uart_manager.c (hunk 2 above)
Comment (yevgenfs) [reply Src/lib/UL/uart_manager/uart_manager.c]: Done

File: Src/stm32f4xx_it.c (hunk 3 above)
Comment (yevgenfs) [reply Src/stm32f4xx_it.c:205]: Done

File: Src/lib/UL/uart_manager/uart_manager.c (hunk 1 above)
Comment (yevgenfs) [reply Src/lib/UL/uart_manager/uart_manager.c]: Done

File: Src/lib/UL/io/io_i2c/io_i2c.c [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+e_io_i2c_err_t send_to_i2c(led_num_t led_num, led_ctrl_t led_state)
+{
+    return e_io_i2c_err_ok;
+}
+
+e_io_i2c_err_t io_i2c_run(void)
+{
+    return e_io_i2c_err_ok;
```
Comment (oleh-plotnikov) [inline Src/lib/UL/io/io_i2c/io_i2c.c]: Please, change to e_io_i2c_err_not_impemented

File: Src/lib/UL/io/io_uart/io_uart.c [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        return e_io_uart_err_invalid_argument;
+    }
+    return e_io_uart_err_ok;
+}
+
+e_io_uart_err_t receive_from_uart()
+{
+    if (uart_receive(&uart, receive_message,
+    RECEIVE_MESSAGE_LENGTH) != e_uart_err_ok)
```
Comment (oleh-plotnikov) [inline Src/lib/UL/io/io_uart/io_uart.c]: missed whitespaces, use small shifting to make it readable

File: Src/lib/UL/io/io_uart/io_uart.c [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            return e_io_uart_err_ok;
+        }
+        return e_io_uart_err_not_found;
+    }
+}
+
+void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart)
+{
+    if (convert_receive_package(receive_message, &receive_package, RECEIVE_MESSAGE_LENGTH) == e_io_uart_err_ok)
```
Comment (oleh-plotnikov) [inline Src/lib/UL/io/io_uart/io_uart.c]: Don't use weight functions in Interrupt callback. Use some variable as flag to identify new received value

**Review by oleh-plotnikov — CHANGES_REQUESTED** [review]

File: Src/lib/UL/io/io_i2c/io_i2c.c (hunk 4 above)
Comment (yevgenfs) [reply Src/lib/UL/io/io_i2c/io_i2c.c]: Done

File: Src/lib/UL/io/io_uart/io_uart.c (hunk 5 above)
Comment (yevgenfs) [reply Src/lib/UL/io/io_uart/io_uart.c]: Done

File: Src/lib/UL/io/io_uart/io_uart.c (hunk 6 above)
Comment (yevgenfs) [reply Src/lib/UL/io/io_uart/io_uart.c]: Done

File: Src/lib/UL/uart_manager/uart_manager.c [hunk 7]
```diff
@@ -34,22 +37,22 @@ e_uart_manager_err_t uart_manager_deinit()
 
 e_uart_manager_err_t convert_header(led_num_t led_num)
 {
-    package.header.led_num = (uint8_t)led_num;
+    send_package.send_header.led_num = (uint8_t)led_num;
```
Comment (oleh-plotnikov) [inline Src/lib/UL/uart_manager/uart_manager.c]: If you defined struct as send_package seems you don't need to call field as send_header, just rename it to "header"

File: Src/lib/UL/uart_manager/uart_manager.c [hunk 8]
```diff
@@ -70,19 +73,72 @@ e_uart_manager_err_t send_to_uart(led_num_t led_num, led_ctrl_t led_state)
...
+//    {
+        if (receive_from_uart() == e_uart_manager_err_ok)
+        {
+            set_spinner_period_ms(get_receive_payload_period_ms());
+        }
+    //}
+    return e_uart_manager_err_ok;
+}
+
+//void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart)
```
Comment (oleh-plotnikov) [inline Src/lib/UL/uart_manager/uart_manager.c]: Remove if it is not necessary

**Review by oleh-plotnikov — APPROVED** [review]
Little bit tricky but okey🙂, let's discuss it later

**Pull request merged**